python/cucim/src/cucim/skimage/morphology/convex_hull.py

```diff
@@ -0,0 +1,315 @@
+]
+
+
+@cp.memoize(for_each_device=True)
+def _offsets_diamond(ndim):
+    offsets = np.zeros((2 * ndim, ndim))
+    for vertex, (axis, offset) in enumerate(product(range(ndim), (-0.5, 0.5))):
+        offsets[vertex, axis] = offset
+    return cp.asarray(offsets)
+
+
+@cp.memoize(for_each_device=True)
+def get_coords_in_hull_kernel(coord_dtype, float_dtype, ndim):
+    """Keep this kernel for n-dimensional support as the raw_moments kernels
+    currently only support 2D and 3D data.
+    """
+    coord_dtype = cp.dtype(coord_dtype)
+    float_dtype = cp.dtype(float_dtype)
+    uint_t = (
+        "unsigned int" if coord_dtype.itemsize <= 4 else "unsigned long long"
+    )
+    float_t = "float" if float_dtype.itemsize <= 4 else "double"
+
+    source = """
+      if (image[i]) {
+        convex_image = true;
+      } else {"""
+    source += _unravel_loop_index("image", ndim, uint_t=uint_t)
+    source += """
+        bool in_hull = true;
+        int n_hull_equations = hull_equations.shape()[0];
+        for (int i_eq = 0; i_eq < n_hull_equations; i_eq++) {"""
+    source += f"""
+          {float_t} v = 0.0;"""
+    for d in range(ndim):
+        source += f"""
+          v += hull_equations[i_eq * {ndim + 1} + {d}] * in_coord[{d}];"""
+    source += f"""
+          v += hull_equations[i_eq * {ndim + 1} + {ndim}];"""
+    source += """
+          if (v > tol) {
+            in_hull = false;
+            break;
+          }
+        }
+        convex_image = in_hull;
+      }\n"""
+    inputs = (
+        f"raw bool image, raw {float_dtype.name} hull_equations, float64 tol"
+    )
+    outputs = "bool convex_image"
+    name = f"cucim_convex_hull_{ndim}d_{coord_dtype.char}_{float_dtype.char}"
+    return cp.ElementwiseKernel(inputs, outputs, source, name=name)
+
+
+def convex_hull_image(
+    image,
+    offset_coordinates=True,
+    tolerance=1e-10,
+    include_borders=True,
+    *,
+    omit_empty_coords_check=False,
+    float64_computation=True,
+    cpu_fallback_threshold=None,
+):
+    """Compute the convex hull image of a binary image.
+
+    The convex hull is the set of pixels included in the smallest convex
+    polygon that surround all white pixels in the input image.
+
+    Parameters
+    ----------
+    image : array
+        Binary input image. This array is cast to bool before processing.
+    offset_coordinates : bool, optional
+        If ``True``, a pixel at coordinate, e.g., (4, 7) will be represented
+        by coordinates (3.5, 7), (4.5, 7), (4, 6.5), and (4, 7.5). This adds
+        some "extent" to a pixel when computing the hull.
+    tolerance : float, optional
+        Tolerance when determining whether a point is inside the hull. Due
+        to numerical floating point errors, a tolerance of 0 can result in
+        some points erroneously being classified as being outside the hull.
+    include_borders: bool, optional
+        If ``False``, vertices/edges are excluded from the final hull mask.
```

Comment:
Since `include_borders=False` case raises the NotImplementedError, it might be good to document this in the docstring.